### PR TITLE
Migrate sync_versions from an API call to a task

### PR DESCRIPTION
**Pull request by ericholscher**
This should stop the API from timing out,
since we aren’t running all this logic in process.
This is most important for projects with lots of versions,
because this logic takes a long time to run and times out the web processes.
It also eats up our API processes,
causing other issues with our API because of blocking.

To test this, you can use the `pull` management command on a project you have checked out locally:

```
inv -e docker.manage 'pull time-test’
```

### CODE REVIEW

**Review by humitos — COMMENTED** [review]
Looks good to me! I left one comment in particular about a potential edge case and two `if` block added that I think are not needed.

**Comment by ericholscher** [comment]
I need to fix up this test, but would love a quick review on the basic idea here.

File: readthedocs/projects/tasks.py [hunk 1]
```diff
@@ -250,19 +251,21 @@ def sync_versions_api(self, version_repo):
...
+            log.info('Remote versions: branches=%s tags=%s', branches, tags)
 
         if all([
             version_repo.supports_tags,
             not self.project.has_feature(Feature.SKIP_SYNC_TAGS)
         ]):
-            tags = tags or version_repo.tags
+            # Will be an empty list if we called lsremote and had no tags returned
+            if tags is None:
+                tags = version_repo.tags
```
[the remark this answers is not in the document]
Comment (ericholscher) [reply readthedocs/projects/tasks.py:268]: No, the comment above explains it. This was causing an error when the feature flag was `True` but returned an empty list.

File: readthedocs/projects/tasks.py [hunk 2]
```diff
@@ -234,13 +235,13 @@ def sync_repo(self, environment):
         )
         version_repo = self.get_vcs_repo(environment)
         version_repo.update()
-        self.sync_versions_api(version_repo)
+        self.sync_versions_task(version_repo)
```
[the remark this answers is not in the document]
Comment (ericholscher) [reply readthedocs/projects/tasks.py:238]: When would the build make use of a version that was created locally? For `stable` builds? If we're already running a build, the version we're running against should already be created.

**Comment by ericholscher** [comment]
I think this is a solid PR, but needs a bunch of test refactoring. I'll jump on it soonish hopefully.

File: readthedocs/projects/tasks.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/projects/tasks.py:268]: Ah, OK! Makes sense. The empty list is valid and we want to keep it that value. Gotcha!

File: readthedocs/projects/tasks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/projects/tasks.py:237]: We already have a version at this point https://github.com/readthedocs/readthedocs.org/blob/466312ea2df09839006fccf666c160d3ed725c17/readthedocs/projects/tasks.py#L238-L239

**Comment by stsewd** [comment]
I have updated this PR

- Split argument into tags and branches data
- Update with the latest changes from `sync_versions`
- Make it backwards compatible (the api endpoint still works)

Still need to update the tests and test it locally.

File: readthedocs/projects/tasks.py [hunk 3]
```diff
@@ -257,39 +257,54 @@ def sync_versions_api(self, version_repo):
...
 
-        try:
-            api_v2.project(self.project.pk).sync_versions.post(
-                version_post_data,
-            )
-        except HttpClientError:
-            log.exception('Sync Versions Exception')
-        except Exception:
-            log.exception('Unknown Sync Versions Exception')
+        from readthedocs.builds import tasks as build_tasks
```
Comment (stsewd) [inline readthedocs/projects/tasks.py:300]: circular imports everywhere!

Comment (stsewd) [reply readthedocs/projects/tasks.py:300]: We could avoid it by moving `send_build_status` to this file as well.

File: readthedocs/builds/tasks.py [hunk 4]
```diff
@@ -206,3 +216,98 @@ def delete_inactive_external_versions(limit=200, days=30 * 3):
...
+    queue='web'
+)
+def sync_versions_task(project_pk, tags_data, branches_data, **kwargs):
+    """
+    Sync the version data in the repo (on the build server).
+
+    Version data in the repo is synced with what we have in the database.
+
+    :param tags_data: List of dictionaries with ``verbose_name`` and ``identifier``.
+    :param branches_data: Same as ``tags_data`` but for branches.
```
Comment (stsewd) [inline readthedocs/builds/tasks.py:233]: I think we should refactor these to be a list of named tuples, I think celery should do fine serializing them, but for another PR anyway...

**Comment by stsewd** [comment]
Looks like this is working fine locally, we could also put it under a feature flag if we want, since the api is still working.

**Comment by stsewd** [comment]
Looks like Eric can't review this PR since he opened it :D

File: readthedocs/projects/tasks.py (hunk 3 above)
Comment (stsewd) [reply readthedocs/projects/tasks.py:300]: https://github.com/readthedocs/readthedocs.org/pull/7764

File: readthedocs/builds/tasks.py (hunk 4 above)
Comment (ericholscher) [reply readthedocs/builds/tasks.py:233]: This should probably be in a `TODO` in the code, then :)

**Review by ericholscher — COMMENTED** [review]
This looks like a solid change 👍 

I think a feature flag would be a good idea, so we can deploy it without much risk. Having it be backwards compat is a really good idea :)

File: readthedocs/api/v2/views/model_views.py [hunk 5]
```diff
@@ -181,47 +163,33 @@ def canonical_url(self, request, **kwargs):
...
-            if 'branches' in data:
-                ret_set = sync_versions_to_db(
-                    project=project,
-                    versions=data['branches'],
-                    type=BRANCH,
-                )
-                added_versions.update(ret_set)
-            deleted_versions = delete_versions_from_db(project, data)
-            deleted_active_versions = get_deleted_active_versions(project, data)
+            added_versions, deleted_versions = sync_versions_task(
```
Comment (humitos) [inline readthedocs/api/v2/views/model_views.py:188]:
```suggestion
            # Calling the task synchronically to keep backward compatibility
            added_versions, deleted_versions = sync_versions_task(
```

File: readthedocs/projects/tasks.py [hunk 6]
```diff
@@ -234,18 +234,18 @@ def sync_repo(self, environment):
         )
         version_repo = self.get_vcs_repo(environment)
         version_repo.update()
-        self.sync_versions_api(version_repo)
+        self.sync_versions_task(version_repo)
         identifier = getattr(self, 'commit', None) or self.version.identifier
         version_repo.checkout(identifier)
 
-    def sync_versions_api(self, version_repo):
+    def sync_versions_task(self, version_repo):
```
Comment (humitos) [inline readthedocs/projects/tasks.py:241]: Maybe we can call this `trigger_sync_versions` or `trigger_sync_versions_task`, to avoid having the same name for this method than for the _real_ task.

File: readthedocs/builds/tasks.py [hunk 7]
```diff
@@ -206,3 +216,98 @@ def delete_inactive_external_versions(limit=200, days=30 * 3):
...
+@app.task(
+    max_retries=1,
+    default_retry_delay=60,
+    queue='web'
+)
+def sync_versions_task(project_pk, tags_data, branches_data, **kwargs):
+    """
+    Sync the version data in the repo (on the build server).
+
+    Version data in the repo is synced with what we have in the database.
```
Comment (humitos) [inline readthedocs/builds/tasks.py:230]:
```suggestion
    Sync the version data in the repo (from build server) into our database.

	Creates new Version objects for those tags/branches that we don't have tracked in our database and deletes Version objects for tags/branches that don't exists anymore in the repository.
```

I found the docstring a little ambiguous and I'm suggesting a change.

**Review by humitos — APPROVED** [review]
Looks good to me!
